src/anndata/_core/merge.py

```diff
@@ -1147,50 +1191,57 @@ def make_dask_col_from_extension_dtype(
-    with maybe_open_h5(base_path_or_zarr_group, elem_name) as f:
-        maybe_chunk_size = get_chunksize(read_elem_lazy(f))
-        chunk_size = (
-            compute_chunk_layout_for_axis_size(
-                1000 if maybe_chunk_size is None else maybe_chunk_size[0], col.shape[0]
-            ),
-        )
-
-    def get_chunk(block_info=None):
-        # reopening is important to get around h5py's unserializable lock in processes
+    if (
+        base_path_or_zarr_group is not None and elem_name is not None
+    ):  # lazy, backed by store
+        dims = col.dims
+        coords = col.coords.copy()
         with maybe_open_h5(base_path_or_zarr_group, elem_name) as f:
-            v = read_elem_lazy(f)
-            variable = xr.Variable(
-                data=xr.core.indexing.LazilyIndexedArray(v), dims=dims
-            )
-            data_array = DataArray(
-                variable,
-                coords=coords,
-                dims=dims,
-            )
-            idx = tuple(
-                slice(start, stop) for start, stop in block_info[None]["array-location"]
+            maybe_chunk_size = get_chunksize(read_elem_lazy(f))
+            chunk_size = (
+                compute_chunk_layout_for_axis_size(
+                    1000 if maybe_chunk_size is None else maybe_chunk_size[0],
+                    col.shape[0],
+                ),
             )
-            chunk = np.array(data_array.data[idx].array)
-        return chunk
 
-    if col.dtype in ("category", "string") or use_only_object_dtype:
-        dtype = "object"
-    else:
-        dtype = col.dtype.numpy_dtype
-    return da.map_blocks(
-        get_chunk,
-        chunks=chunk_size,
-        meta=np.array([], dtype=dtype),
-        dtype=dtype,
-    )
+        def get_chunk(block_info=None):
+            # reopening is important to get around h5py's unserializable lock in processes
+            with maybe_open_h5(base_path_or_zarr_group, elem_name) as f:
+                v = read_elem_lazy(f)
+                variable = xr.Variable(
+                    data=xr.core.indexing.LazilyIndexedArray(v), dims=dims
+                )
+                data_array = XArray(
+                    variable,
+                    coords=coords,
+                    dims=dims,
+                )
+                idx = tuple(
+                    slice(start, stop)
+                    for start, stop in block_info[None]["array-location"]
+                )
+                chunk = np.array(data_array.data[idx].array)
+            return chunk
+
+        if col.dtype == "category" or col.dtype == "string" or use_only_object_dtype:  # noqa PLR1714
+            dtype = "object"
+        else:
+            dtype = col.dtype.numpy_dtype
+        return da.map_blocks(
+            get_chunk,
+            chunks=chunk_size,
+            meta=np.array([], dtype=dtype),
+            dtype=dtype,
+        )
+    else:  # in-memory
+        return da.from_array(col.values, chunks=-1)
```

Comment:
```suggestion
    return da.from_array(col.values, chunks=-1)
```

Reply:
Further, why create a dask array on top of the values? You want the returned value to still be lazy here?  I think this should be some sort of flag then on `concat` - with backed datasets, we have basically no choice until something better is figured out but for this case, this is a very strong choice.

Reply:
I added a new argument to `concat` that controls this.